inference-engine/src/readers/onnx_reader/onnx_model_validator.cpp

```diff
@@ -0,0 +1,235 @@
+            GRAPH, OPSET_IMPORT, METADATA_PROPS, TRAINING_INFO
+        };
+
+        const auto is_allowed = [&decoded_field](const Field field) {
+            return field == static_cast<Field>(decoded_field);
+        };
+
+        return std::any_of(std::begin(allowed_fields), std::end(allowed_fields), is_allowed);
+    }
+
+    bool is_correct_wire_type_for_field(const PbKey& decoded_key) {
+        static const std::map<Field, WireType> combinations = {
+            {IR_VERSION, VARINT},
+            {PRODUCER_NAME, LENGTH_DELIMITED},
+            {PRODUCER_VERSION, LENGTH_DELIMITED},
+            {DOMAIN_, LENGTH_DELIMITED},
+            {MODEL_VERSION, VARINT},
+            {DOC_STRING, LENGTH_DELIMITED},
+            {GRAPH, LENGTH_DELIMITED},
+            {OPSET_IMPORT, LENGTH_DELIMITED},
+            {METADATA_PROPS, LENGTH_DELIMITED},
+            {TRAINING_INFO, LENGTH_DELIMITED},
+        };
+
+        return combinations.at(static_cast<Field>(decoded_key.first)) == static_cast<WireType>(decoded_key.second);
+    }
+
+    /**
+     * Only 7 bits in each component of a varint count in this algorithm. The components form
+     * a decoded number when they are concatenated bitwise in a reverse order. For example:
+     * bytes = [b1, b2, b3, b4]
+     * varint = b4 ++ b3 ++ b2 ++ b1  <== only 7 bits of each byte should be extracted before concat
+     *
+     *             b1         b2
+     * bytes = [00101100, 00000010]
+     *             b2         b1
+     * varint = 0000010 ++ 0101100 = 100101100 => decimal: 300
+     * Each consecutive varint byte needs to be left shifted "7 x its position in the vector"
+     * and bitwise added to the accumulator afterwards.
+     */
+    uint32_t varint_bytes_to_number(const std::vector<char>& bytes) {
+        uint32_t accumulator = 0u;
+
+        for (size_t i = 0; i < bytes.size(); ++i) {
+            uint32_t b = bytes[i];
+            b <<= 7 * i;
+            accumulator |= b;
+        }
+
+        return accumulator;
+    }
+
+    uint32_t decode_varint(std::istream& model) {
+        std::vector<char> bytes;
+        bytes.reserve(4);
+
+        char key_component = 0;
+        model.get(key_component);
+
+        while (key_component & 0x80) {
+            // drop the most significant bit
+            const char component = key_component & ~0x80;
+            bytes.push_back(component);
+            model.get(key_component);
+        }
+        // add the last byte - the one with MSB off
+        bytes.push_back(key_component);
+
+        return varint_bytes_to_number(bytes);
+    }
+
+    PbKey decode_key(const char key) {
+        // 3 least significant bits
+        const char wire_type = key & 0b111;
+        // remaining bits
+        const char field_number = key >> 3;
+        return {field_number, wire_type};
+    }
+
+    ONNXField decode_next_field(std::istream& model) {
+        char key = 0;
+        model.get(key);
```

Comment:
The function name suggest that you want to decode next field but key is always 0.
Why is that?

Reply:
The key is retrieved from the stream using `std::istream::get`, zero is just an initial value for this variable.